format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -30,30 +32,131 @@
+import java.util.ArrayList;
 import java.util.Iterator;
+import java.util.List;
+import java.util.TreeMap;
 import javax.annotation.Nullable;
 
-/**
- * Delimited text format that tracks which file each record was read from.
- */
+/** Delimited text format that tracks which file each record was read from. */
 public class PathTrackingDelimitedInputFormat extends PathTrackingInputFormat {
+
   static final String DELIMITER = "delimiter";
+  static final String ENABLE_QUOTES_VALUE = "enable_quotes_value";
   static final String SKIP_HEADER = "skip_header";
 
+  private static final String QUOTE = "\"";
+  private static final char QUOTE_CHAR = '\"';
+
+  /**
+   * Split the delimited string based on the delimiter. The delimiter should not contain any quotes.
+   * The method will behave like this: 1. if there is no quote, it will behave same as {@link
+   * String#split(String)} 2. if there are quotes in the string, the method will find pairs of
+   * quotes, content within each pair of quotes will not get splitted even if there is delimiter in
+   * that. For example, if string is a."b.c"."d.e.f" and delimiter is '.', it will get split into
+   * [a, b.c, d.e.f]. if string is "val1.val2", then it will not get splitted since the '.' is
+   * within pair of quotes. if string is "val1.val2"", the quote right after val2 will be kept as
+   * part of the value, so [val1, val2"].
+   *
+   * @param delimitedString the string to split
+   * @param delimiter the separtor
+   * @return a list of splits of the original string
+   */
+  @VisibleForTesting
+  static List<String> splitQuotesString(String delimitedString, String delimiter) {
+    // use a tree map so we can easily look up if a delimiter is within a pair of quotes
+    // the map has key of index and value to the index of paired quote. The map key is the index of
+    // the starting and the ending quote and the value is the index of the paired quote. An
+    // un-enclosed quote will trigger the exception and fail the pipeline. Example:
+    // pos: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
+    //      " | a | , | b | " | , | " | c | , | d | "  |
+    // key-value pairs in the map:
+    // 0 -> 4
+    // 4 -> 0
+    // 6 -> 10
+    // 10 -> 6
+    TreeMap<Integer, Integer> quotesPos = new TreeMap<>();
+    int firstQuotePos = -1;
+    for (int i = 0; i < delimitedString.length(); i++) {
+      if (delimitedString.charAt(i) != QUOTE_CHAR) {
+        continue;
+      }
+
+      if (firstQuotePos == -1) {
+        firstQuotePos = i;
+      } else {
+        quotesPos.put(i, firstQuotePos);
+        quotesPos.put(firstQuotePos, i);
+        firstQuotePos = -1;
+      }
+    }
+
+    List<String> result = new ArrayList<>();
+    StringBuilder split = new StringBuilder();
+    for (int i = 0; i < delimitedString.length(); i++) {
+      // if the length is not enough for the delimiter, just add it to split
+      if (i + delimiter.length() > delimitedString.length()) {
```

Comment:
do it like
```
if (i + delimiter.length() > delimitedString.length()) {
  split.append(...)
  continue;
}
...
```
to avoid the huge nesting block